File: examples/2015-day-01/clojure/test/aoc_test.clj
(ns aoc-test
  (:require [aoc :as aoc]
            [clojure.test :refer [deftest is]]))

(def part-one-cases
  [["(())" 0]
   ["()()" 0]
   ["(((" 3]
   ["(()(()(" 3]
   ["))(((((" 3]
   ["())" -1]
   ["))(" -1]
   [")))" -3]
   [")())())" -3]])

(def part-two-cases
  [[")" 1]
   ["()())" 5]])

(deftest part-one-solves-the-examples
  (doseq [[input expected] part-one-cases]
    (is (= expected (aoc/solve-part-one input))
        (str "failed on input " (pr-str input)))))

(deftest part-two-solves-the-examples
  (doseq [[input expected] part-two-cases]
    (is (= expected (aoc/solve-part-two input))
        (str "failed on input " (pr-str input)))))

File: examples/2015-day-01/scala/test/aoc.test.scala
class AocSuite extends munit.FunSuite:
  /** Checks one solution function against every example in the puzzle statement. */
  private def checkSolution(solve: String => Long, cases: Seq[(String, Long)]): Unit =
    for (input, expected) <- cases do
      assertEquals(solve(input), expected, s"failed on input $input")

  private val partOneCases = Seq(
    "(())" -> 0L,
    "()()" -> 0L,
    "(((" -> 3L,
    "(()(()(" -> 3L,
    "))(((((" -> 3L,
    "())" -> -1L,
    "))(" -> -1L,
    ")))" -> -3L,
    ")())())" -> -3L
  )

  private val partTwoCases = Seq(")" -> 1L, "()())" -> 5L)

  test("part one solves the examples"):
    checkSolution(solvePartOne, partOneCases)

  test("part two solves the examples"):
    checkSolution(solvePartTwo, partTwoCases)
